music_assistant/server/providers/apple_music/__init__.py

```diff
@@ -0,0 +1,687 @@
+        response = await self._get_data(endpoint)
+        return [self._parse_album(album) for album in response["data"] if album["id"]]
+
+    async def get_artist_toptracks(self, prov_artist_id) -> list[Track]:
+        """Get a list of 10 most popular tracks for the given artist."""
+        endpoint = f"catalog/{self._storefront}/artists/{prov_artist_id}/view/top-songs"
+        response = await self._get_data(endpoint)
+        return [self._parse_track(track) for track in response["data"] if track["id"]]
+
+    async def library_add(self, item: MediaItemType):
+        """Add item to library."""
+        raise NotImplementedError("Not implemented!")
+
+    async def library_remove(self, prov_item_id, media_type: MediaType):
+        """Remove item from library."""
+        raise NotImplementedError("Not implemented!")
+
+    async def add_playlist_tracks(self, prov_playlist_id: str, prov_track_ids: list[str]):
+        """Add track(s) to playlist."""
+        raise NotImplementedError("Not implemented!")
+
+    async def remove_playlist_tracks(
+        self, prov_playlist_id: str, positions_to_remove: tuple[int, ...]
+    ) -> None:
+        """Remove track(s) from playlist."""
+        raise NotImplementedError("Not implemented!")
+
+    async def get_similar_tracks(self, prov_track_id, limit=25) -> list[Track]:
+        """Retrieve a dynamic list of tracks based on the provided item."""
+        # Note, Apple music does not have an official endpoint for similar tracks.
+        # We will use the next-tracks endpoint to get a list of tracks that are similar to the
+        # provided track. However, Apple music only provides 2 tracks at a time, so we will
+        # need to call the endpoint multiple times. Therefore, set a limit to 6 to prevent
+        # flooding the apple music api.
+        limit = 6
+        endpoint = f"me/stations/next-tracks/ra.{prov_track_id}"
+        found_tracks = []
+        while len(found_tracks) < limit:
+            response = await self._post_data(endpoint, include="artists")
+            if not response or "data" not in response:
+                break
+            for track in response["data"]:
+                if track and track["id"]:
+                    found_tracks.append(self._parse_track(track))
+        return found_tracks
+
+    async def get_stream_details(self, item_id: str) -> StreamDetails:
+        """Return the content details for the given track when it will be streamed."""
+        stream_metadata = await self._fetch_song_stream_metadata(item_id)
+        license_url = stream_metadata["hls-key-server-url"]
+        stream_url, uri = self._parse_stream_url_and_uri(stream_metadata["assets"])
+        key_id = base64.b64decode(uri.split(",")[1])
+        return StreamDetails(
+            item_id=item_id,
+            provider=self.instance_id,
+            audio_format=AudioFormat(
+                content_type=ContentType.MP4,
+            ),
```

Comment:
this is a bit dangerous as MP4 is a container that could contain multiple codecs. Our ContenType enum is in fact not flexible enough to describe that as it more described the container format and not the actual codec. If you are 100% sure the codec is always AAC (and not HE-AAC), you could set the contenttype to AAC but otherwise just set it to UNKNOWN and it will be auto detected when the stream starts

Reply:
`AAC` did not seem to work on all music providers for some reason. I set it to `UNKNOWN` which seem to work great. It does detect the input to be aac anyway though:
`Detected (input) content type: aac (aac)`